File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-2.swift
class Solution {
    func findMin(_ nums: [Int]) -> Int {
          // O(logn)
            var result = nums[0]
            var left = 0
            var right = nums.count - 1
            
            while left <= right {
                // edge case
                if nums[left] < nums[right] {
                    result = min(result, nums[left])
                    break
                }
                let mid = (left + right) / 2
                result = min(result, nums[mid])
                if nums[mid] >= nums[left]{
                    // search the right sorted portion
                    left = mid + 1
                }else{
                    right = mid - 1
                }
                
            }
            
            return result

    }
}

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-2.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:

        L = 0 
        R = len(nums) - 1

        while L <= R:
            mid = ( R + L ) // 2

            if target == nums[mid]:
                return mid

            # left side is sorted
            if nums[L] <= nums[mid]:
                if target > nums[mid] or target < nums[L]:
                    L = mid + 1 
                else: 
                    R = mid - 1 
            # right portion 
            else: 
                if target < nums[mid] or target > nums[R]:
                    R = mid - 1
                else: 
                    L = mid + 1 
        return -1 
                



            

        